File: src/co/nums/intellij/aem/htl/file/HtlLanguageSubstitutor.java
package co.nums.intellij.aem.htl.file;

import co.nums.intellij.aem.constants.JcrConstants;
import co.nums.intellij.aem.htl.HtlLanguage;
import com.intellij.ide.highlighter.HtmlFileType;
import com.intellij.lang.Language;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.LanguageSubstitutor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class HtlLanguageSubstitutor extends LanguageSubstitutor {

	@Nullable
	@Override
	public Language getLanguage(@NotNull VirtualFile file, @NotNull Project project) {
		if (isHtml(file) && isInJcrRootDirectory(file)) {
			return HtlLanguage.INSTANCE;
		}
		return null;
	}

	private static boolean isHtml(@NotNull VirtualFile file) {
		return file.getFileType() == HtmlFileType.INSTANCE;
	}

	private static boolean isInJcrRootDirectory(@NotNull VirtualFile file) {
		VirtualFile parent = file.getParent();
		while (parent != null) {
			if (JcrConstants.JCR_ROOT_DIRECTORY_NAME.equals(parent.getName())) {
				return true;
			}
			parent = parent.getParent();
		}
		return false;
	}

}

File: src/co/nums/intellij/aem/htl/file/HtlLanguageSubstitutor.kt
package co.nums.intellij.aem.htl.file

import co.nums.intellij.aem.constants.JcrConstants
import co.nums.intellij.aem.htl.HtlLanguage
import com.intellij.ide.highlighter.HtmlFileType
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.LanguageSubstitutor

class HtlLanguageSubstitutor : LanguageSubstitutor() {

    override fun getLanguage(file: VirtualFile, project: Project) =
            if (isHtml(file) && isInJcrRootDirectory(file)) HtlLanguage
            else null

    private fun isHtml(file: VirtualFile) = file.fileType === HtmlFileType.INSTANCE

    private fun isInJcrRootDirectory(file: VirtualFile): Boolean {
        var parent: VirtualFile? = file.parent
        while (parent != null) {
            if (parent.name == JcrConstants.JCR_ROOT_DIRECTORY_NAME) {
                return true
            }
            parent = parent.parent
        }
        return false
    }

}
